### DISCUSSION TITLE
[back] 차단 된 유저

### DISCUSSION THREAD

**Discussion by kyu-baek** [3DPong/transcendence#126] [Q&A]
### 서브젝트에 차단 관련 사항은 아래의 메세지가 전부 입니다

> The user should be able to block other users. This way, they will see no more messages from the account they blocked.

단체 채팅방에서 차단된 유저 메세지는 차단한 당사자한테 안뜨게만 처리해야 하는 걸까요?
채팅방을 만드는 owner가 여러 사람을 초대하는데, 초대 유저들끼리의 차단 관계를 다 계산해서 생성을 막게 하는 건 오히려 이상 한 것 같습니다.
그럼 owner 를 차단하고 있는 사람을 초대 할 수 없게는 해야하는 건지... 등등 명확하지가 않아서 질문드립니다.

우선 **will see no more messages** 부분을 보면 메시지만 보이지 않는다는 말처럼 보이기는 합니다. 

그런데 초대나, 생성은 다 가능한테 메세지를 안보이게 하는 건 아래에 같은 설계가 요구될 것 같습니다.
1. 차단한 시점부터 이후에 보내는 메세지를 안 보이게 하는 거라면 -> 그럼 차단 시간도 필요해지고, message 로그 api 로 보내줄때 일일이 비교해가며 체크해야 합니다.
2. 소켓 통신때도 차단 관계를 체크하면서 emit 할지 말지를 비교해야 합니다.

##
#### _a 유저가 b 유저를 차단하면,_
```

1. dm
  a. 우선 b유저는 a 에게 dm 메세지를 더이상 보낼 수 없게 된다. -> 당신은 a에게 차단 되었습니다! 에러 날리기
  b.  b 입장에서는 자기가 차단 당한 줄도 모르고 (카톡 처럼?) 일단 메세지를 보내고 db에 저장할 순 있지만, a에게 띄워주지 않는다.

2. 단체 채팅방
 a. a 유저와 b 유저가 함께 속해 있는 단체 채팅 방에서는 b는 메세지를 보낼 수 있고, a는 b 의 것만 안 받는다. or b (...) 으로 내용만 가리고 받게 한다.
 b. 이미 함께 포함된 채팅방에서는 영향을 주지 않게 한다. -> 마음에 안들면 a 가 나가기 하면 됨

3. 새로 만들기
 a. b가 오너일 땐, a를 초대할 수 없다.
 b. a가 오너일 땐, b를 초대할 수 있다?
c. 초대이던, 추가이던, 채팅방을 파는 건 뭐든 가능. 하지만 메세지만 못받음..


```

**Comment by csstring** [comment]
저희 초창기 회의때 카카오톡 으로 차단에 대한 테스트를 진행을 했던 결과, 차단 유저가 1대1 대화 메세지를 보내면 차단 되지만 채팅방에서는 차단이 되지 않는 것을 보고 카카오톡을 예시 삼아 초기 설정을 잡았던 걸로 기억합니다.
그래서 dm 부분만 차단 구현을 해놔도 될거 같습니다.

dm에서 차단기능을 일단 구현해 놓으면 채팅방에서 차단기능이 되든 안되는 평가라는 기준에서 둘다 디펜스가 될거 같고, 초기 설계 목적도 부합하는거 같습니다

**Comment by parktest0325** [comment]
전체채팅에서도 차단 기능을 만들면 좋을거같긴한데, 저도 dm차단 기능만 만들어도 디펜스가 될것같고, 전체 채팅 차단기능은 좀더 복잡해질거같다는 생각이 드네요.
 
dm 차단은 기능이 구현됐다는걸 보여주기 쉽게 차단된 유저에게 알려주는게 좋을것같고, 
차단된걸 알고있으면 메시지는 저장되지 않아도 괜찮을것같아요 => 이건 그냥 선택사항인듯 싶습니다 

만약 dm 이외의 방에서 차단을 막으려면, 처음 채팅방 들어갈때 :channelId/log API에서도 차단관계 확인 후 그 메시지를 필터링 해야할것같네요. 
그렇게 구현되면 20개 기준으로 메시지 페이징을 처리하고있어서 메시지 검색할때 in으로 차단리스트를 제외하고 20개를 뽑아와야 할것같아요.

**Comment by kimminkyeu** [comment]
저는 개발관점을 떠나서 보면 dm은 B안이 좋다고 생각합니다.
그 이유는 첫째, 차단된 사용자가 내가 차단되었음을 인지할 수 있다면, 새로운 계정을 파서 친추를 걸고 계속 dm을 보내는 방식이 가능해지기 때문입니다.
찾아보니 이런 방식 방지하는 걸 "Shadow-Ban" 이라고 부르나보네요
참고 링크 : https://influencermarketinghub.com/instagram-shadowban/

결론 : 
1. DM = B안->차단된 사람은 "내가 차단되었다"라고 인지를 못하게 하도록 하는 방식
2. 채팅방 = 디펜스만 충분하다면 채팅방 차단은 구현하지 않아도 된다고 보나, 만약 구현한다면 C방식 (초대이던, 추가이던, 채팅방을 파는 건 뭐든 가능하나 메시지만 못받는 방식)
